File: testing/tests/TS-717/test_ts_717.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:trackstate/data/repositories/trackstate_repository.dart';
import 'package:trackstate/data/services/local_workspace_onboarding_service.dart';
import 'package:trackstate/data/services/local_workspace_onboarding_service_io.dart';
import 'package:trackstate/data/services/workspace_profile_service.dart';

import '../../fixtures/workspace_onboarding_screen_fixture.dart';
import 'support/ts717_ready_workspace_fixture.dart';

const String _ticketKey = 'TS-717';
const String _ticketSummary =
    'Local folder inspection recognizes a usable TrackState repository as ready';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    SharedPreferences.setMockInitialValues(const <String, Object>{});
  });

  testWidgets(
    'TS-717 local folder inspection keeps a ready repository openable without writes',
    (tester) async {
      final result = <String, Object?>{
        'ticket': _ticketKey,
        'ticket_summary': _ticketSummary,
        'environment': 'flutter test',
        'os': Platform.operatingSystem,
        'steps': <Map<String, Object?>>[],
        'human_verification': <Map<String, Object?>>[],
      };

      final semantics = tester.ensureSemantics();
      final fixture = await tester.runAsync(Ts717ReadyWorkspaceFixture.create);
      if (fixture == null) {
        semantics.dispose();
        throw StateError('TS-717 fixture creation did not complete.');
      }
      addTearDown(() async {
        await tester.runAsync(fixture.dispose);
      });

      try {
        final service = const LocalGitWorkspaceOnboardingService();
        final beforeSnapshot = await tester.runAsync(fixture.captureSnapshot);
        if (beforeSnapshot == null) {
          throw StateError(
            'TS-717 pre-open repository snapshot did not complete.',
          );
        }

        result['repository_path'] = fixture.repositoryPath;
        result['workspace_folder_name'] = fixture.workspaceFolderName;
        result['before_head_revision'] = beforeSnapshot.headRevision;
        result['before_worktree_status'] = beforeSnapshot.worktreeStatusLines;
        result['before_files'] = beforeSnapshot.files;

        final inspection = await tester.runAsync(
          () => service.inspectFolder(fixture.repositoryPath),
        );
        if (inspection == null) {
          throw StateError('TS-717 inspection did not complete.');
        }

        result['inspection'] = <String, Object?>{
          'state': inspection.state.name,
          'message': inspection.message,
          'folderPath': inspection.folderPath,
          'suggestedWorkspaceName': inspection.suggestedWorkspaceName,
          'suggestedWriteBranch': inspection.suggestedWriteBranch,
          'detectedWriteBranch': inspection.detectedWriteBranch,
          'hasGitRepository': inspection.hasGitRepository,
          'canOpen': inspection.canOpen,
        };

        if (inspection.state != LocalWorkspaceInspectionState.readyToOpen ||
            !inspection.canOpen ||
            inspection.folderPath != fixture.repositoryPath ||
            inspection.suggestedWorkspaceName != fixture.workspaceFolderName ||
            inspection.suggestedWriteBranch !=
                Ts717ReadyWorkspaceFixture.defaultBranch ||
            inspection.detectedWriteBranch !=
                Ts717ReadyWorkspaceFixture.defaultBranch ||
            !inspection.hasGitRepository) {
          throw AssertionError(
            'Step 3 failed: LocalWorkspaceInspectionService did not classify the prepared repository as a ready local workspace.\n'
            'Observed state: ${inspection.state.name}\n'
            'Observed message: ${inspection.message}\n'
            'Observed folderPath: ${inspection.folderPath}\n'
            'Observed suggestedWorkspaceName: ${inspection.suggestedWorkspaceName}\n'
            'Observed suggestedWriteBranch: ${inspection.suggestedWriteBranch}\n'
            'Observed detectedWriteBranch: ${inspection.detectedWriteBranch}\n'
            'Observed hasGitRepository: ${inspection.hasGitRepository}\n'
            'Observed canOpen: ${inspection.canOpen}',
          );
        }

        final pickerInvocations = <Map<String, String?>>[];
        final openedRepositories = <String>[];
        final onboardingService = _RecordingLocalWorkspaceOnboardingService(
          inspection,
        );
        final workspaceProfileService =
            SharedPreferencesWorkspaceProfileService(
              now: () => DateTime.utc(2026, 5, 14, 16, 30),
            );

        final screen = await launchWorkspaceOnboardingFixture(
          tester,
          workspaceProfileService: workspaceProfileService,
          localWorkspaceOnboardingService: onboardingService,
          workspaceDirectoryPicker:
              ({String? confirmButtonText, String? initialDirectory}) async {
                pickerInvocations.add(<String, String?>{
                  'confirmButtonText': confirmButtonText,
                  'initialDirectory': initialDirectory,
                });
                return fixture.repositoryPath;
              },
          openLocalRepository:
              ({
                required String repositoryPath,
                required String defaultBranch,
                required String writeBranch,
              }) async {
                openedRepositories.add(
                  '$repositoryPath@$defaultBranch@$writeBranch',
                );
                return const DemoTrackStateRepository();
              },
        );

        try {
          final initialState = screen.captureState();
          final initialVisibleTexts = initialState.visibleTexts;
          result['initial_visible_texts'] = initialVisibleTexts;
          if (!initialVisibleTexts.contains('Add workspace') ||
              !initialVisibleTexts.contains('Open existing folder')) {
            throw AssertionError(
              'Step 1 failed: the onboarding screen did not render the expected Add workspace entry point before choosing a local folder.\n'
              'Observed visible texts: ${_formatList(initialVisibleTexts)}',
            );
          }

          await screen.chooseExistingFolder();

          if (pickerInvocations.length != 1) {
            throw AssertionError(
              'Step 2 failed: selecting "Open existing folder" did not invoke the DirectoryPickerAdapter exactly once.\n'
              'Observed picker invocations: ${pickerInvocations.length}\n'
              'Observed invocations: ${jsonEncode(pickerInvocations)}',
            );
          }

          final pickerInvocation = pickerInvocations.single;
          result['picker_invocation'] = pickerInvocation;
          if (pickerInvocation['confirmButtonText'] !=
              'Choose existing folder') {
            throw AssertionError(
              'Step 2 failed: the DirectoryPickerAdapter was not called with the expected open-existing confirmation label.\n'
              'Observed confirmButtonText: ${pickerInvocation['confirmButtonText'] ?? '<null>'}\n'
              'Observed initialDirectory: ${pickerInvocation['initialDirectory'] ?? '<null>'}',
            );
          }
          result['inspected_folder_paths'] =
              onboardingService.inspectedFolderPaths;
          if (onboardingService.inspectedFolderPaths.isEmpty ||
              onboardingService.inspectedFolderPaths.any(
                (path) => path != fixture.repositoryPath,
              )) {
            throw AssertionError(
              'Step 3 failed: the onboarding UI did not inspect the selected local folder path.\n'
              'Observed inspected folder paths: ${_formatList(onboardingService.inspectedFolderPaths)}\n'
              'Expected selected folder path: ${fixture.repositoryPath}',
            );
          }
          _recordStep(
            result,
            step: 1,
            status: 'passed',
            action:
                'Launch the app and select "Open existing folder" on the onboarding screen.',
            observed:
                'visible_texts=${_formatList(initialVisibleTexts)}; picker_invoked=true',
          );
          _recordStep(
            result,
            step: 2,
            status: 'passed',
            action:
                'Use the DirectoryPickerAdapter to select the prepared directory.',
            observed:
                'confirmButtonText=${pickerInvocation['confirmButtonText']}; selectedFolder=${fixture.repositoryPath}',
          );

          final localSelectionState = screen.captureState();
          final visibleTexts = localSelectionState.visibleTexts;
          final interactiveLabels =
              localSelectionState.interactiveSemanticsLabels;
          final nameValue = localSelectionState.localWorkspaceNameValue;
          final writeBranchValue = localSelectionState.localWriteBranchValue;
          final submitLabel = localSelectionState.primaryActionLabel;
          result['visible_texts'] = visibleTexts;
          result['interactive_semantics_labels'] = interactiveLabels;
          result['workspace_name_value'] = nameValue;
          result['write_branch_value'] = writeBranchValue;
          result['submit_label'] = submitLabel;
          result['submit_enabled'] = localSelectionState.isPrimaryActionEnabled;

          final requiredTexts = <String>[
            'Ready to open',
            'This folder already contains a committed TrackState workspace and is ready to open.',
            'Selected folder',
            fixture.repositoryPath,
            'Workspace details',
            'Workspace name',
            'Write Branch',
            'Open workspace',
          ];
          final missingTexts = requiredTexts
              .where((text) => !visibleTexts.contains(text))
              .toList(growable: false);
          if (missingTexts.isNotEmpty ||
              nameValue != fixture.workspaceFolderName ||
              writeBranchValue != Ts717ReadyWorkspaceFixture.defaultBranch ||
              submitLabel != 'Open workspace' ||
              !localSelectionState.isPrimaryActionEnabled) {
            throw AssertionError(
              'Step 3 failed: the ready repository selection did not advance to the expected Workspace details state with an enabled Open action.\n'
              'Missing visible texts: ${_formatList(missingTexts)}\n'
              'Observed visible texts: ${_formatList(visibleTexts)}\n'
              'Observed workspace name: $nameValue\n'
              'Observed write branch: $writeBranchValue\n'
              'Observed submit label: $submitLabel\n'
              'Observed submit enabled: ${localSelectionState.isPrimaryActionEnabled}',
            );
          }
          _recordStep(
            result,
            step: 3,
            status: 'passed',
            action:
                'Observe the LocalWorkspaceInspectionService output and the visible UI state.',
            observed:
                'inspection_state=${inspection.state.name}; inspected_folder=${_formatList(onboardingService.inspectedFolderPaths)}; status=Ready to open; selected_folder=${fixture.repositoryPath}; workspace_name=$nameValue; write_branch=$writeBranchValue; submit_label=$submitLabel; submit_enabled=${localSelectionState.isPrimaryActionEnabled}',
          );

          final requiredSemanticFragments = <String>[
            'Open existing folder',
            'Initialize folder',
            'Change folder',
            'Workspace name',
            'Write Branch',
            'Open workspace',
          ];
          final missingSemanticFragments = requiredSemanticFragments
              .where(
                (fragment) =>
                    !interactiveLabels.any((label) => label.contains(fragment)),
              )
              .toList(growable: false);
          if (missingSemanticFragments.isNotEmpty ||
              interactiveLabels.any((label) => label.trim().isEmpty)) {
            throw AssertionError(
              'Step 4 failed: one or more interactive onboarding elements did not expose a non-empty semantics label.\n'
              'Missing semantic fragments: ${_formatList(missingSemanticFragments)}\n'
              'Observed interactive semantics labels: ${_formatList(interactiveLabels)}',
            );
          }
          _recordStep(
            result,
            step: 4,
            status: 'passed',
            action:
                'Verify that interactive elements have non-empty Semantics labels (AC6).',
            observed:
                'interactive_semantics_labels=${_formatList(interactiveLabels)}',
          );

          await screen.submit();

          final afterSnapshot = await tester.runAsync(fixture.captureSnapshot);
          if (afterSnapshot == null) {
            throw StateError(
              'TS-717 post-open repository snapshot did not complete.',
            );
          }
          result['after_head_revision'] = afterSnapshot.headRevision;
          result['after_worktree_status'] = afterSnapshot.worktreeStatusLines;
          result['after_files'] = afterSnapshot.files;
          result['opened_repositories'] = openedRepositories;
          final postOpenState = screen.captureState();
          result['dashboard_visible'] = postOpenState.isDashboardVisible;

          if (_singleOrNull(openedRepositories) !=
                  '${fixture.repositoryPath}@main@main' ||
              afterSnapshot.headRevision != beforeSnapshot.headRevision ||
              !_listEquals(
                afterSnapshot.worktreeStatusLines,
                beforeSnapshot.worktreeStatusLines,
              ) ||
              !_mapEquals(afterSnapshot.files, beforeSnapshot.files) ||
              !postOpenState.isDashboardVisible) {
            throw AssertionError(
              'Step 3 failed: opening the ready repository changed files on disk or called the local repository loader with unexpected arguments.\n'
              'Observed opened repositories: ${_formatList(openedRepositories)}\n'
              'Observed head revision before/after: ${beforeSnapshot.headRevision} -> ${afterSnapshot.headRevision}\n'
              'Observed worktree status before: ${_formatList(beforeSnapshot.worktreeStatusLines)}\n'
              'Observed worktree status after: ${_formatList(afterSnapshot.worktreeStatusLines)}\n'
              'Observed files changed: ${!_mapEquals(afterSnapshot.files, beforeSnapshot.files)}\n'
              'Observed dashboard visible: ${postOpenState.isDashboardVisible}',
            );
          }

          _recordHumanVerification(
            result,
            check:
                'Observed the visible onboarding copy a user sees after choosing the folder, including the ready status, selected folder path, workspace details heading, and enabled Open workspace action.',
            observed:
                'visible_texts=${_formatList(visibleTexts)}; interactive_semantics_labels=${_formatList(interactiveLabels)}',
          );
          _recordHumanVerification(
            result,
            check:
                'Pressed "Open workspace" and verified the dashboard became visible while the selected repository stayed byte-for-byte unchanged on disk.',
            observed:
                'dashboard_visible=true; opened_repositories=${_formatList(openedRepositories)}; head_revision=${afterSnapshot.headRevision}; worktree_status=${_formatList(afterSnapshot.worktreeStatusLines)}; file_manifest_unchanged=true',
          );

          print('TS-717-OBSERVATION:${jsonEncode(result)}');
        } finally {
          screen.dispose();
        }
      } finally {
        semantics.dispose();
      }
    },
    timeout: const Timeout(Duration(seconds: 60)),
  );
}

String? _singleOrNull(List<String> values) {
  if (values.length != 1) {
    return null;
  }
  return values.single;
}

void _recordStep(
  Map<String, Object?> result, {
  required int step,
  required String status,
  required String action,
  required String observed,
}) {
  final steps =
      result.putIfAbsent('steps', () => <Map<String, Object?>>[])
          as List<Map<String, Object?>>;
  steps.add(<String, Object?>{
    'step': step,
    'status': status,
    'action': action,
    'observed': observed,
  });
}

void _recordHumanVerification(
  Map<String, Object?> result, {
  required String check,
  required String observed,
}) {
  final checks =
      result.putIfAbsent('human_verification', () => <Map<String, Object?>>[])
          as List<Map<String, Object?>>;
  checks.add(<String, Object?>{'check': check, 'observed': observed});
}

String _formatList(List<Object?> values) {
  if (values.isEmpty) {
    return '<empty>';
  }
  return values.map((value) => value.toString()).join(' | ');
}

bool _listEquals(List<String> left, List<String> right) {
  if (left.length != right.length) {
    return false;
  }
  for (var index = 0; index < left.length; index += 1) {
    if (left[index] != right[index]) {
      return false;
    }
  }
  return true;
}

bool _mapEquals(Map<String, String> left, Map<String, String> right) {
  if (left.length != right.length) {
    return false;
  }
  for (final entry in left.entries) {
    if (right[entry.key] != entry.value) {
      return false;
    }
  }
  return true;
}

class _RecordingLocalWorkspaceOnboardingService
    implements LocalWorkspaceOnboardingService {
  _RecordingLocalWorkspaceOnboardingService(this.inspection);

  final LocalWorkspaceInspection inspection;
  final List<String> inspectedFolderPaths = <String>[];

  @override
  Future<LocalWorkspaceInspection> inspectFolder(String folderPath) async {
    inspectedFolderPaths.add(folderPath);
    return inspection;
  }

  @override
  Future<LocalWorkspaceSetupResult> initializeFolder({
    required LocalWorkspaceInspection inspection,
    required String workspaceName,
    required String writeBranch,
  }) {
    throw UnimplementedError(
      'TS-717 only exercises the ready-to-open onboarding path.',
    );
  }
}

File: testing/tests/TS-717/test_ts_717.py
from __future__ import annotations

import json
import re
import subprocess
import sys
import traceback
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[3]
if str(REPO_ROOT) not in sys.path:
    sys.path.insert(0, str(REPO_ROOT))

TICKET_KEY = "TS-717"
TICKET_SUMMARY = (
    "Local folder inspection recognizes a usable TrackState repository as ready"
)
OUTPUTS_DIR = REPO_ROOT / "outputs"
JIRA_COMMENT_PATH = OUTPUTS_DIR / "jira_comment.md"
PR_BODY_PATH = OUTPUTS_DIR / "pr_body.md"
RESPONSE_PATH = OUTPUTS_DIR / "response.md"
RESULT_PATH = OUTPUTS_DIR / "test_automation_result.json"
BUG_DESCRIPTION_PATH = OUTPUTS_DIR / "bug_description.md"
TEST_FILE_PATH = "testing/tests/TS-717/test_ts_717.dart"
RUN_COMMAND = "flutter test testing/tests/TS-717/test_ts_717.dart --reporter expanded"
REQUEST_STEPS = [
    'Launch the app and select "Open existing folder" on the Onboarding screen.',
    "Use the DirectoryPickerAdapter to select the prepared directory.",
    "Observe the LocalWorkspaceInspectionService output and UI state.",
    "Verify that interactive elements have non-empty Semantics labels (AC6).",
]


def main() -> None:
    OUTPUTS_DIR.mkdir(parents=True, exist_ok=True)
    try:
        completed = subprocess.run(
            ["flutter", "test", TEST_FILE_PATH, "--reporter", "expanded"],
            cwd=REPO_ROOT,
            capture_output=True,
            text=True,
            check=False,
        )
        combined_output = _combine_output(completed.stdout, completed.stderr)
        observation = _extract_json_line(combined_output, "TS-717-OBSERVATION:")

        if completed.returncode != 0:
            raise AssertionError(_extract_error_message(combined_output))
        if observation is None:
            raise AssertionError(
                "Flutter test passed but did not emit the expected TS-717 observation payload."
            )

        _write_pass_outputs(
            observation=observation,
            combined_output=combined_output,
        )
    except Exception as error:
        combined_output = locals().get("combined_output", "") or ""
        _write_failure_outputs(error=error, combined_output=combined_output)
        raise


def _write_pass_outputs(*, observation: dict[str, object], combined_output: str) -> None:
    BUG_DESCRIPTION_PATH.unlink(missing_ok=True)
    RESULT_PATH.write_text(
        json.dumps(
            {
                "status": "passed",
                "passed": 1,
                "failed": 0,
                "skipped": 0,
                "summary": "1 passed, 0 failed",
            }
        )
        + "\n",
        encoding="utf-8",
    )

    inspection = observation.get("inspection")
    inspection_payload = inspection if isinstance(inspection, dict) else {}
    picker_payload = observation.get("picker_invocation")
    picker_invocation = picker_payload if isinstance(picker_payload, dict) else {}
    visible_texts = _as_list(observation.get("visible_texts"))
    semantics_labels = _as_list(observation.get("interactive_semantics_labels"))
    opened_repositories = _as_list(observation.get("opened_repositories"))
    human_checks = observation.get("human_verification")
    human_verifications = human_checks if isinstance(human_checks, list) else []
    visible_texts_text = ", ".join(visible_texts) or "<none>"
    semantics_text = ", ".join(semantics_labels) or "<none>"
    opened_repositories_text = ", ".join(opened_repositories) or "<none>"
    output_excerpt = _compact_text(combined_output)

    inspection_state = _as_text(inspection_payload.get("state"))
    inspection_message = _as_text(inspection_payload.get("message"))
    selected_folder = _as_text(inspection_payload.get("folderPath"))
    workspace_name = _as_text(observation.get("workspace_name_value"))
    write_branch = _as_text(observation.get("write_branch_value"))
    submit_label = _as_text(observation.get("submit_label"))
    picker_confirm = _as_text(picker_invocation.get("confirmButtonText"))
    dashboard_visible = observation.get("dashboard_visible") is True
    before_head = _as_text(observation.get("before_head_revision"))
    after_head = _as_text(observation.get("after_head_revision"))

    jira_lines = [
        "h3. Test Automation Result",
        "",
        "*Status:* ✅ PASSED",
        f"*Test Case:* {TICKET_KEY} — {TICKET_SUMMARY}",
        "",
        "h4. What was automated",
        (
            f"* Step 1: Launched the onboarding flow and activated "
            f"{{{{Open existing folder}}}} from the visible {{Add workspace}} screen."
        ),
        (
            f"* Step 2: Verified the DirectoryPickerAdapter callback ran with "
            f"{{code}}confirmButtonText={picker_confirm}{{code}} and returned the prepared local repository "
            f"{{code}}{selected_folder}{{code}}."
        ),
        (
            f"* Step 3: Verified {{code}}LocalWorkspaceInspectionService.inspectFolder(...){{code}} "
            f"reported {{code}}{inspection_state}{{code}}, rendered the visible ready state, prefilled "
            f"workspace name {{code}}{workspace_name}{{code}}, prefilled write branch {{code}}{write_branch}{{code}}, "
            f"enabled the {{code}}{submit_label}{{code}} action, opened the workspace, and left the repository unchanged on disk."
        ),
        (
            f"* Step 4: Verified interactive controls exposed non-empty semantics labels, including "
            f"{jira_inline('Open existing folder')}, {jira_inline('Change folder')}, "
            f"{jira_inline('Workspace name')}, {jira_inline('Write branch')}, and {jira_inline('Open workspace')}."
        ),
        "",
        "h4. Human-style verification",
        (
            f"* Observed the user-facing copy after folder selection: status {jira_inline('Ready to open')}, "
            f"message {jira_inline(inspection_message)}, selected folder {jira_inline(selected_folder)}, "
            f"heading {jira_inline('Workspace details')}, and enabled action {jira_inline(submit_label)}."
        ),
        (
            f"* Pressed {jira_inline(submit_label)} and observed the visible workspace shell continue to "
            f"{jira_inline('Dashboard')} while the repository stayed unchanged "
            f"(HEAD {jira_inline(before_head)} -> {jira_inline(after_head)}, opened repository call "
            f"{jira_inline(opened_repositories_text)})."
        ),
    ]

    for check in human_verifications:
        if not isinstance(check, dict):
            continue
        jira_lines.append(
            f"* {jira_inline(_as_text(check.get('check')))} — observed {jira_inline(_as_text(check.get('observed')))}"
        )

    jira_lines.extend(
        [
            "",
            "h4. Result",
            "* Step 1 passed: the onboarding screen exposed the expected open-existing entry point.",
            "* Step 2 passed: the directory picker callback was used for the prepared local repository.",
            "* Step 3 passed: the prepared repository was recognized as ready, the Workspace details surface showed the expected values, the Open action was enabled, and opening the workspace did not modify the selected folder on disk.",
            "* Step 4 passed: interactive elements kept non-empty semantics labels.",
            f"* Observed visible texts: {{code}}{visible_texts_text}{{code}}",
            f"* Observed interactive semantics labels: {{code}}{semantics_text}{{code}}",
            f"* Dashboard visible after opening: {{code}}{str(dashboard_visible).lower()}{{code}}",
            "* The observed behavior matched the expected result.",
            "",
            "h4. Run command",
            "{code:bash}",
            RUN_COMMAND,
            "{code}",
            "",
            "h4. Observed output excerpt",
            "{code}",
            output_excerpt,
            "{code}",
        ]
    )

    markdown_lines = [
        "## Test Automation Result",
        "",
        "**Status:** ✅ PASSED",
        f"**Test Case:** {TICKET_KEY} — {TICKET_SUMMARY}",
        "",
        "## What was automated",
        '- Launched the onboarding flow and selected `Open existing folder` from the visible `Add workspace` screen.',
        (
            f"- Verified the DirectoryPickerAdapter callback ran with "
            f"`confirmButtonText={picker_confirm}` and returned the prepared repository `{selected_folder}`."
        ),
        (
            f"- Verified `LocalWorkspaceInspectionService.inspectFolder(...)` reported `{inspection_state}`, "
            f"rendered the ready state, prefilled workspace name `{workspace_name}`, prefilled write branch "
            f"`{write_branch}`, and enabled the `{submit_label}` action."
        ),
        (
            f"- Pressed `{submit_label}`, verified the dashboard became visible, and confirmed the selected "
            f"repository stayed unchanged on disk (`HEAD {before_head} -> {after_head}`, loader call `{opened_repositories_text}`)."
        ),
        (
            f"- Verified interactive controls exposed non-empty semantics labels: `{semantics_text}`."
        ),
        "",
        "## Human-style verification",
        (
            f"- Observed the user-facing ready state copy in place: `Ready to open`, "
            f"`{inspection_message}`, selected folder `{selected_folder}`, heading `Workspace details`, "
            f"and enabled action `{submit_label}`."
        ),
        (
            f"- Observed the post-open experience a user would see: the app continued to `Dashboard` without changing the selected folder's files."
        ),
    ]

    for check in human_verifications:
        if not isinstance(check, dict):
            continue
        markdown_lines.append(
            f"- `{_as_text(check.get('check'))}` — observed `{_as_text(check.get('observed'))}`"
        )

    markdown_lines.extend(
        [
            "",
            "## Result",
            "- Step 1 passed: the onboarding screen exposed the expected open-existing action.",
            "- Step 2 passed: the directory picker callback was used for the prepared repository.",
            "- Step 3 passed: the repository was recognized as ready, Workspace details rendered the expected values, the Open action was enabled, and opening the workspace did not modify the selected folder.",
            "- Step 4 passed: interactive elements kept non-empty semantics labels.",
            "- The observed behavior matched the expected result.",
            "",
            "## How to run",
            "```bash",
            RUN_COMMAND,
            "```",
        ]
    )

    JIRA_COMMENT_PATH.write_text("\n".join(jira_lines) + "\n", encoding="utf-8")
    PR_BODY_PATH.write_text("\n".join(markdown_lines) + "\n", encoding="utf-8")
    RESPONSE_PATH.write_text("\n".join(markdown_lines) + "\n", encoding="utf-8")


def _write_failure_outputs(*, error: Exception, combined_output: str) -> None:
    error_message = f"{type(error).__name__}: {error}"
    RESULT_PATH.write_text(
        json.dumps(
            {
                "status": "failed",
                "passed": 0,
                "failed": 1,
                "skipped": 0,
                "summary": "0 passed, 1 failed",
                "error": error_message,
            }
        )
        + "\n",
        encoding="utf-8",
    )

    annotated_steps = []
    if "Step 1 failed:" in combined_output:
        annotated_steps.extend(
            [
                f"1. ❌ {REQUEST_STEPS[0]} Observed: the onboarding screen did not expose the expected open-existing entry point or did not invoke it successfully. See the assertion output below.",
                f"2. ⏭️ {REQUEST_STEPS[1]} Not reached because step 1 failed.",
                f"3. ⏭️ {REQUEST_STEPS[2]} Not reached because step 1 failed.",
                f"4. ⏭️ {REQUEST_STEPS[3]} Not reached because step 1 failed.",
            ]
        )
    elif "Step 2 failed:" in combined_output:
        annotated_steps.extend(
            [
                f"1. ✅ {REQUEST_STEPS[0]} Observed: the onboarding screen launched and the open-existing action was activated.",
                f"2. ❌ {REQUEST_STEPS[1]} Observed: the DirectoryPickerAdapter callback parameters or invocation count did not match the expected open-existing flow. See the assertion output below.",
                f"3. ⏭️ {REQUEST_STEPS[2]} Not reached because step 2 failed.",
                f"4. ⏭️ {REQUEST_STEPS[3]} Not reached because step 2 failed.",
            ]
        )
    elif "Step 3 failed:" in combined_output:
        annotated_steps.extend(
            [
                f"1. ✅ {REQUEST_STEPS[0]} Observed: the onboarding screen launched and the open-existing action was activated.",
                f"2. ✅ {REQUEST_STEPS[1]} Observed: the directory picker callback ran for the prepared local repository.",
                f"3. ❌ {REQUEST_STEPS[2]} Observed: the prepared repository was not recognized or rendered as the expected ready-to-open workspace state, or opening it changed disk state unexpectedly. See the assertion output below.",
                f"4. ⏭️ {REQUEST_STEPS[3]} Not reached because step 3 failed.",
            ]
        )
    elif "Step 4 failed:" in combined_output:
        annotated_steps.extend(
            [
                f"1. ✅ {REQUEST_STEPS[0]} Observed: the onboarding screen launched and the open-existing action was activated.",
                f"2. ✅ {REQUEST_STEPS[1]} Observed: the directory picker callback ran for the prepared local repository.",
                f"3. ✅ {REQUEST_STEPS[2]} Observed: the repository reached the ready UI state.",
                f"4. ❌ {REQUEST_STEPS[3]} Observed: one or more interactive elements did not keep a non-empty semantics label. See the assertion output below.",
            ]
        )
    else:
        annotated_steps.extend(
            [
                f"1. ❌ {REQUEST_STEPS[0]} Observed: the failure occurred before the test reached a step-specific assertion boundary.",
                f"2. ⏭️ {REQUEST_STEPS[1]} Not reached.",
                f"3. ⏭️ {REQUEST_STEPS[2]} Not reached.",
                f"4. ⏭️ {REQUEST_STEPS[3]} Not reached.",
            ]
        )

    failure_excerpt = combined_output.strip() or error_message
    jira_lines = [
        "h3. Test Automation Result",
        "",
        "*Status:* ❌ FAILED",
        f"*Test Case:* {TICKET_KEY} — {TICKET_SUMMARY}",
        "",
        "h4. Failure summary",
        f"* Exact error: {{code}}{error_message}{{code}}",
        "",
        "h4. Step-by-step reproduction with observations",
        *[f"* {line}" for line in annotated_steps],
        "",
        "h4. Actual vs Expected",
        "* Expected: selecting a prepared committed local TrackState repository should classify it as ready, render the Workspace details step with the Open action enabled, keep interactive semantics labels non-empty, and leave the repository unchanged on disk.",
        f"* Actual: {{code}}{_compact_text(failure_excerpt)}{{code}}",
        "",
        "h4. Environment",
        "* URL: N/A (local flutter widget test)",
        "* Browser: N/A",
        "* OS: Linux",
        f"* Run command: {{code}}{RUN_COMMAND}{{code}}",
        "",
        "h4. Exact assertion output / stack trace",
        "{code}",
        failure_excerpt,
        "{code}",
    ]

    markdown_lines = [
        "## Test Automation Result",
        "",
        "**Status:** ❌ FAILED",
        f"**Test Case:** {TICKET_KEY} — {TICKET_SUMMARY}",
        "",
        "## Failure summary",
        f"**Exact error:** `{error_message}`",
        "",
        "## Step-by-step reproduction with observations",
        *[f"1. {line[3:]}" if line.startswith('1.') else line for line in []],
    ]
    markdown_lines.extend(f"- {line}" for line in annotated_steps)
    markdown_lines.extend(
        [
            "",
            "## Actual vs Expected",
            "- Expected: selecting a prepared committed local TrackState repository should classify it as ready, render the Workspace details step with the Open action enabled, keep interactive semantics labels non-empty, and leave the repository unchanged on disk.",
            f"- Actual: `{_compact_text(failure_excerpt)}`",
            "",
            "## Environment",
            "- URL: N/A (local flutter widget test)",
            "- Browser: N/A",
            "- OS: Linux",
            f"- Run command: `{RUN_COMMAND}`",
            "",
            "## Exact assertion output / stack trace",
            "```text",
            failure_excerpt,
            "```",
        ]
    )

    bug_lines = [
        f"# {TICKET_KEY} automated test failure",
        "",
        f"## Summary",
        f"The automated reproduction for **{TICKET_SUMMARY}** failed.",
        "",
        "## Exact steps to reproduce",
        *annotated_steps,
        "",
        "## Exact error message or assertion failure",
        "```text",
        failure_excerpt,
        "```",
        "",
        "## Actual vs Expected",
        "- Expected: choosing the prepared committed local TrackState repository should classify it as ready, render the Workspace details state with an enabled Open action, keep interactive semantics labels non-empty, and avoid modifying files on disk.",
        f"- Actual: {_compact_text(failure_excerpt)}",
        "",
        "## Environment details",
        "- URL: N/A (local flutter widget test)",
        "- Browser: N/A",
        "- OS: Linux",
        f"- Run command: `{RUN_COMMAND}`",
        f"- Test file: `{TEST_FILE_PATH}`",
        "",
        "## Screenshots or logs",
        "- Screenshot: not applicable for this flutter widget test run.",
        "```text",
        failure_excerpt,
        "```",
    ]

    JIRA_COMMENT_PATH.write_text("\n".join(jira_lines) + "\n", encoding="utf-8")
    PR_BODY_PATH.write_text("\n".join(markdown_lines) + "\n", encoding="utf-8")
    RESPONSE_PATH.write_text("\n".join(markdown_lines) + "\n", encoding="utf-8")
    BUG_DESCRIPTION_PATH.write_text("\n".join(bug_lines) + "\n", encoding="utf-8")


def _extract_json_line(output: str, prefix: str) -> dict[str, object] | None:
    match = re.search(rf"{re.escape(prefix)}(\{{.*\}})", output)
    if not match:
        return None
    return json.loads(match.group(1))


def _extract_error_message(output: str) -> str:
    step_match = re.search(r"(Step \d+ failed:.*?)(?:\n\s*\n|\n══╡|\Z)", output, re.S)
    if step_match:
        return _compact_text(step_match.group(1))
    human_match = re.search(
        r"(Human-style verification failed:.*?)(?:\n\s*\n|\n══╡|\Z)", output, re.S
    )
    if human_match:
        return _compact_text(human_match.group(1))
    return "AssertionError: flutter test exited with a non-zero status"


def _combine_output(stdout: str, stderr: str) -> str:
    stdout = stdout.strip()
    stderr = stderr.strip()
    if stdout and stderr:
        return f"{stdout}\n{stderr}"
    return stdout or stderr


def _as_text(value: object | None) -> str:
    if value is None:
        return "<none>"
    if isinstance(value, bool):
        return "true" if value else "false"
    return str(value)


def _as_list(value: object | None) -> list[str]:
    if not isinstance(value, list):
        return []
    return [str(item) for item in value]


def _compact_text(text: str) -> str:
    compact = re.sub(r"\s+", " ", text).strip()
    if len(compact) <= 1800:
        return compact
    return compact[:1800] + "…"


def jira_inline(text: str) -> str:
    escaped = text.replace("{", r"\{").replace("}", r"\}")
    return f"{{{{{escaped}}}}}"


if __name__ == "__main__":
    try:
        main()
    except Exception:
        traceback.print_exc()
        raise
